_posts/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node.md

```diff
@@ -0,0 +1,467 @@
+nodes.
+
+![Metal3 Simulation
+Workflow](/assets/2024-10-24-Scaling-Kubernetes-with-Metal3-on-Fake-Node/metal3-simulation-workflow.jpg)
+
+### Step 1: Setting Up the environment
+
+As you may have known, a typical Metal3 workflow requires several components:
+bootstrap Kubernetes cluster, possible external networks, bare-metal nodes, etc.
+As we are working on simulating the environment, we will start with a newly
+spawned Ubuntu VM, create a cluster with minikube, add networks with libvirt,
+and so on (If you're familiar with Metal3's dev-env, this step is similar to
+what script
+[01](https://github.com/metal3-io/metal3-dev-env/blob/main/01_prepare_host.sh),
+[02](https://github.com/metal3-io/metal3-dev-env/blob/main/02_configure_host.sh)
+and a part of
+[03](https://github.com/metal3-io/metal3-dev-env/blob/main/03_launch_mgmt_cluster.sh)
+do). We will not discuss this part, but you can find the related setup from
+[this
+script](https://github.com/Nordix/metal3-clusterapi-docs/blob/main/Support/Multitenancy/Scalability-with-Fake-Nodes/vm-setup.sh)
+if interested.
+
+__Note__: If you intend to follow along, note that going to 1000 nodes requires
+a large environment and will take a long time. In our setup, we had a VM with 24
+cores and 32GB of RAM, of which we assigned 14 cores and 20GB of RAM to the
+minikube VM, and the process took roughly 48 hours. If your environment is less
+powerful, consider reducing the nodes you want to provision. Something like 100
+nodes will require minimal resources and time while still being impressive.
+
+### Step 2: Install BMO and Ironic
+
+In Metal3's typical workflow, we usually rely on Kustomize to install Ironic and
+BMO. Kustomize helps us define configurations for Kubernetes resources, making
+it easier to customize and deploy services. However, our current Kustomize
+overlay for Metal3 configures only a single Ironic instance. This setup works
+well for smaller environments, but it becomes a bottleneck when scaling up and
+handling thousands of nodes.
+
+That's where Ironic's __special mode__ comes into play. Ironic has the ability
+to run __multiple Ironic conductors__ while sharing the same database. The best
+part? Workload balancing between conductors happens automatically, which means
+that no matter which Ironic conductor receives a request, the load is evenly
+distributed across all conductors, ensuring efficient provisioning. Achieving
+this requires separating __ironic conductor__ from the database, which allows us
+to scale up the conductor part. Each __conductor__ will have its own
+`PROVISIONING_IP,` hence the need to have a specialized `configMap.`
+
+We used [Helm](https://helm.sh/) for this purpose. In our Helm chart, the
+__Ironic conductor__ container and __HTTP server (httpd)__ container are
+separated into a new pod, and the rest of the ironic package (mostly
+`MariaDB`-ironic database) stays in another pod. A list of PROVISIONING_IPs is
```

Comment:
This looks quite strange rendered when one part of the word is in ticks and the other outside. I would just skip the ticks. They aren't usually used for product names anyway.
```suggestion
MariaDB-ironic database) stays in another pod. A list of PROVISIONING_IPs is
```